Feature: Toggle Item Restriction based on Container Type

searica opened the pull request:
This PR adds a toggle-able feature that restricts the dynamic containers it adds to only allow placing items of a specific type within the dynamic container. These restrictions can be toggled on/off in-game safely and restrictions are ignored whenever a container is loading from ZDO so that items that may have been placed inside it when it was not restricted are not accidentally deleted when loading the container with restrictions enabled.

I have tested that this feature works as intended in single-player when used alongside QuickStackStore and AzuAutoStore, even if non-allowed item are present in a container both mods do not stack new items to it if the restrictions are enabled. It also works with FastItemTransfer, AzuHoverStats, ContentsWithin, AzuCraftyBoxes, and CraftFromContainers. 

All test were performing using each method of placing items in containers:
- Ctrl + Click
- Holding interact
- Click and move item to place in container

I was not able to test in a multiplayer setting, so I have not tested for possible bugs when multiple players are using the mod or when MultiUserChest is also present.

Most of the changes to code and functionally is contained within the RestrictContainers.cs file and I have done my best to match the existing code conventions within the mod.

searica committed:
Add RestrictContainers.cs

searica committed:
Create static Patch class handle restricting placement of items in container inventories

searica committed:
Add config entry to toggle container restrictions

searica committed:
Create dictionary of container names and allowed items to pass to RestructContainers since prefab name can't be retrieved from the perspective of the inventory

searica committed:
Fix duplicate implementation warning

searica committed:
Fix imports

searica committed:
Improve consistency of code style

searica committed:
Restore white space lines

searica committed:
Fix error due to disposable method arg

MSchmoecker commented on DynamicStoragePiles/RestrictContainers.cs:
```diff
@@ -0,0 +1,298 @@
+
+        /// <summary>
+        ///    Patch to trigger CanAddItem check when attempting to add item to inventory.
+        ///    If CanAddItem is false then the AddItem method in Valheim will be skipped
+        ///    and return false to indicate the item was not added.
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="item"></param>
+        /// <param name="__result"></param>
+        /// <returns></returns>
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData) })]
+        [HarmonyPriority(Priority.First)]
+        private static bool AddItemPrefix_1(Inventory __instance, ItemDrop.ItemData item, ref bool __result) {
+            if (!CanAddItem(__instance, item)) {
+                __result = false;
+                return __result;
+            }
+            return true;
+        }
+
+        /// <summary>
+        ///    Patch to trigger CanAddItem check when attempting to add item to inventory.
+        ///    If CanAddItem is false then the AddItem method in Valheim will be skipped
+        ///    and return false to indicate the item was not added.
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="item"></param>
+        /// <param name="__result"></param>
+        /// <returns></returns>
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData), typeof(int), typeof(int), typeof(int) })]
+        [HarmonyPriority(Priority.First)]
+        private static bool AddItemPrefix_2(Inventory __instance, ItemDrop.ItemData item, int amount, ref bool __result) {
+            if (!CanAddItem(__instance, item)) {
+                __result = false;
+                return __result;
+            }
+            return true;
+        }
```
```suggestion
        /// <summary>
        ///    Patch to trigger CanAddItem check when attempting to add item to inventory.
        ///    If CanAddItem is false then the AddItem method in Valheim will be skipped
        ///    and return false to indicate the item was not added.
        /// </summary>
        /// <param name="__instance"></param>
        /// <param name="item"></param>
        /// <param name="__result"></param>
        /// <returns></returns>
        [HarmonyPrefix]
        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData) })]
        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData), typeof(int), typeof(int), typeof(int) })]
        [HarmonyPriority(Priority.First)]
        private static bool AddItemPrefix(Inventory __instance, ItemDrop.ItemData item, ref bool __result) {
            if (!CanAddItem(__instance, item)) {
                __result = false;
                return __result;
            }
            return true;
        }
```

The two patches can be merged together

MSchmoecker replied:
Wait, `typeof(Inventory)` is missing here as well. Did I miss a Harmony feature where the type isn't nessesary or was this case catched by a different patch?

searica replied:
There is a Harmony feature where you can declare typeof(Inventory) on the class and it will apply to to all methods within the class so the code takes advantage of that since all the patch methods in the class are typeof(Inventory). Happy to change that if explicitly setting typeof for each method fits the conventions for DynamicStoragePiles better though.

Also thanks for letting me know about merging the patches together, I wasn't sure how Harmony would handle patching multiple methods with different signatures when the patch method used a subset of the params, since I only want to patch 2 out of the 3 AddItem methods in the code. Now that I know they can be merged into one patch I'll do that though as it's a cleaner solution. Thank you for the edits here and in your other suggestions!

MSchmoecker replied:
Oh right, missed the HarmonyPatch at the class level. I rather prefer to explicitly put the type on the methods and not the class. The reason is readability IMO and it's easier to expand the class with a patch to a different class in the future.

searica replied:
Yeah, I recently found that to be the case while working on another mod where I had to patch a variety of different classes to achieve a desired functionality. I've just added the explicit type to the AddItem patch methods and removed the typeof at the class level for RestricContianers.

MSchmoecker commented on DynamicStoragePiles/RestrictContainers.cs:
```diff
@@ -0,0 +1,298 @@
+            Jotunn.Logger.LogDebug($"Item: {item.PrefabName()}");
+            
+            return CanAddItem(__instance, item);
+        }
```
```suggestion
        /// <summary>
        ///     Patch to prevent MoveItemToThis from running if CanAddItem check is false. MoveItemToThis
        ///     will call RemoveItem from the source inventory even in cases where AddItem returns false,
        ///     so the entire method needs to be prevented from running to avoid items being lost when players
        ///     try to place items in containers that do not allow that type of item.
        /// </summary>
        /// <param name="__instance"></param>
        /// <param name="fromInventory"></param>
        /// <param name="item"></param>
        /// <returns></returns>
        [HarmonyPrefix]
        [HarmonyPatch(typeof(Inventory), nameof(Inventory.MoveItemToThis), new[] { typeof(Inventory), typeof(ItemDrop.ItemData) })]
        [HarmonyPatch(typeof(Inventory), nameof(Inventory.MoveItemToThis), new[] { typeof(Inventory), typeof(ItemDrop.ItemData), typeof(int), typeof(int), typeof(int) })]
        [HarmonyPriority(Priority.First)]
        private static bool MoveItemToThisPrefix(Inventory __instance, Inventory fromInventory, ItemDrop.ItemData item) {
            if (__instance == null || fromInventory == null || item == null) { return false; }

            Jotunn.Logger.LogDebug("MoveItemToThisPrefix");
            Jotunn.Logger.LogDebug($"Add to: {__instance.m_name}");
            Jotunn.Logger.LogDebug($"Item: {item.PrefabName()}");
            
            return CanAddItem(__instance, item);
        }
```

The two patches can also be merged together. Also, typeof(Inventory) was missing on the second patch, so it wasn't actually applied.

MSchmoecker commented on DynamicStoragePiles/RestrictContainers.cs:
```diff
@@ -0,0 +1,298 @@
+        [HarmonyPriority(Priority.First)]
+        private static bool RemoveItemPrefix_2(Inventory __instance, ItemDrop.ItemData item) {
+            return ShouldRemoveItem(__instance, item);
+        }
```
```suggestion
        /// <summary>
        ///     Patch to check trigger ShouldRemoveItem check and prevent RemoveItem from running if
        ///     the item was moved during a call to MoveAll and failed to be added to the target
        ///     container because it was restricted and not the allowed item type.
        /// </summary>
        /// <param name="__instance"></param>
        /// <param name="item"></param>
        /// <returns></returns>
        [HarmonyPrefix]
        [HarmonyPatch(typeof(Inventory), nameof(Inventory.RemoveItem), new[] { typeof(ItemDrop.ItemData) })]
        [HarmonyPatch(typeof(Inventory), nameof(Inventory.RemoveItem), new[] { typeof(ItemDrop.ItemData), typeof(int) })]
        [HarmonyPriority(Priority.First)]
        private static bool RemoveItemPrefix(Inventory __instance, ItemDrop.ItemData item) {
            return ShouldRemoveItem(__instance, item);
        }
```

Here the two patches are also the same. Also, `typeof(Inventory)` was missing here as well

MSchmoecker commented on DynamicStoragePiles/RestrictContainers.cs:
```diff
@@ -0,0 +1,298 @@
+        private static string _loadingContainer;
+        private static string _targetContainer;
+        private static string _allowedItem;
+        private static Dictionary<string, string> _allowedItemsByContainer = new Dictionary<string, string>();
```
The second Dictionary seems a bit convoluted, the Plugin allowedItemsByContainer could be made public and accessed from the class.

searica replied:
Yeah, it is a bit convoluted, I had used that approach since I wrote this code in parallel with what I wrote for OdinsFoodBarrels so I tried to make the functionality of RestrictContainers as encapsulated and portable as possible. For this use case making the dictionary from Plugin internal or public is a better solution.

MSchmoecker commented on DynamicStoragePiles/RestrictContainers.cs:
```diff
@@ -0,0 +1,298 @@
+            // Return early if this check is being called while loading a container.
+            // Don't want to delete "non-allowed" items that were put in the container
+            // while the restrictions were disabled.
+            if (!string.IsNullOrEmpty(_loadingContainer) && inventory.m_name == _loadingContainer) {
```
```suggestion
            if (inventory.m_name == _loadingContainer) {
```

the string.IsNullOrEmpty check isn't needed, since two strings are compared

MSchmoecker reviewed (commented):
Looks pretty solid, I only have some minor comments. Thanks a lot!

searica committed:
Remove uneeded IsNullOrEmpty check

Co-authored-by: Maximilian Schmöcker <39767545+MSchmoecker@users.noreply.github.com>

searica committed:
Merge patches with same implementations

Co-authored-by: Maximilian Schmöcker <39767545+MSchmoecker@users.noreply.github.com>

searica committed:
Merge patches with same implementations

Co-authored-by: Maximilian Schmöcker <39767545+MSchmoecker@users.noreply.github.com>

searica committed:
Update DynamicStoragePiles/RestrictContainers.cs

Co-authored-by: Maximilian Schmöcker <39767545+MSchmoecker@users.noreply.github.com>

searica committed:
Make allowedItemsByContainer public

searica committed:
Use public allowedItemsByContainer from Plugin

searica committed:
Add typeof(Inventory) to patch method

searica committed:
Remove typeof from top level class

MSchmoecker commented:
There's one major bug, where switching a restricted item with a non-restricted item causes it to be deleted.

https://github.com/MSchmoecker/Dynamic-Storage-Piles/assets/39767545/24ec1199-2c2f-4c3d-a38d-5b13be7c3e62

When this is resolved and no other issues pop up, I'm going to merging it.

There's some minor cleanup left, like removing debug logging and moving the displayed text to a language file but I'll do that when the rest is done and it's fully merged.

searica committed:
Merge remote-tracking branch 'upstream/master' into feat-restrict-containers

searica committed:
Explicit typeof in patch methods

searica committed:
Remove unused import

searica committed:
Patch to fix item deletion when moving items being containers

searica committed:
Code style

searica committed:
Remove debug logging

searica commented:
Oh damn, thank you for catching that bug! I've just pushed some commits that include a new patch to InventoryGrid.DropItem that appears to fix the bug when I was testing it. I also removed the debug logging and cleaned up the code style a bit more.

MSchmoecker commented:
Everything seems to be working fine now, thanks again!
